WIP: ETCD-295: Automated Backups of Etcd

hasbro17 opened the pull request:
Enhancement proposal for enabling automated backups of etcd in Openshift.
See:
https://issues.redhat.com/browse/ETCD-295
https://issues.redhat.com/browse/ETCD-81
https://issues.redhat.com/browse/OCPBU-252

/hold
For further iteration on outstanding TODOs.

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+To enable automated backups a new singleton CRD can be used to specify the backup schedule and other configuration as well as monitor the status of the backup operations. 
+
+A new controller in the cluster-etcd-operator would then reconcile this CRD to save backups locally per the specified configuration.
```
super nitpick, I think we should make the existing `UpgradeBackupController` more general instead of writing an entirely new controller.

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+        // e.g retain 500mb of backups, or discard backups older than a month
+        // If we want to extend this API in the future we should make retention
+        // a type so we can add more optional fields later
+    RetentionCount int `json: "retentionCount"`
```
if we already have good ideas on other types of retention, maybe add a struct for this? I know it's TP MVP, but let's not introduce something we have to completely change in v1alpha2

deads2k replied:
>if we already have good ideas on other types of retention, maybe add a struct for this? I know it's TP MVP, but let's not introduce something we have to completely change in v1alpha2

Agreed, you want to get close on your first attempt and this looks ripe for discriminated  union.

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+To enforce the specified schedule we can utilize [CronJobs](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) to run backup pods.
+
+**TODO:** If load balancing between control-plane nodes is a requirement then it needs to be determined how we can achieve that with the Job or Pod spec.
```
actually, that's an interesting issue we wouldn't have with a single mounted PVC. 
Let's say I have RetentionCount = 3, how do we know where to find and delete the respective snapshot on which CP node?

jewzaam commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+The procedure also requires gaining a root shell on a control plane
+node. Shell access to OpenShift control plane nodes access is generally
+discouraged due to the potential for affecting the reliability of the
```
Beyond discouraged, it is not possible for hosted control planes in a managed environment where hosted cluster admin does not administer the env in which the hosted control plane (and etcd) is running.

jewzaam commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+- One-time cluster backup can be triggered without requiring a root shell
+- Scheduled backups can be configured after cluster installation
+- Backups are saved locally on the host filesystem of control-plane nodes
```
How does this work for hosted control planes?
1. can the admin in the hosted cluster, who doesn't manage the control plane, initiate this backup?
2. if they can what protections are in place to ensure backups do not impact the hosting node in a way that may negaively impact other control plane pods hosted on the node where the backup is saved?

wking replied:
"Separate ownership" today, doesn't mean it needs to be separate forever, does it?  It seems like it would be possible to move towards a unified approach.  For example, in a hosted control plane world, it's not clear to me [which persona(s)][1] would have which access to backups.  Possibly the Cluster Service Consumer, if we lump backup restoration in as similar to "drive upgrades".  Or possibly the Cluster Service Provider, if we lump backup restoration in under "Has cluster admin management".  But either way, I'd expect etcd backup/restore tooling to live on the management cluster and not leak out to the `Cluster Instance ...` roles.  Which would argue for [a namespaced CRD][2], so a management cluster could have these resources in each of its per-hosted-cluster namespaces.

[1]: https://hypershift-docs.netlify.app/reference/concepts-and-personas/#personas
[2]: https://github.com/openshift/enhancements/pull/1370#discussion_r1163950907

tjungblu replied:
interesting take, I would rather see more separation instead of more unification. For example, we rebased etcd earlier this year and discovered it broke a customer cluster. checkout https://github.com/etcd-io/etcd/issues/15062

We didn't even know that Hypershift was using openshift/etcd as their etcd. They are certainly not using CEO, which makes reusing the backup CRD moot as they would have to implement their own business logic for backups anyway. 

CEO is also currently a singleton in the cluster (don't think we want to change that either), so namespacing that CRD would make little sense in OCP.

jewzaam commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+### Non-Goals
+- Have automated backups enabled by default with cluster installation
+- Save cluster backups to PersistentVolumes or cloud storage
+  - This would be a future enhancement or extension to the API
```
This may be required for adoption in managed openshift and hosted control planes to mitigate possible risks.  For automated backups to be useful for managed it should be usable by the end customer.  But the end customer is not managing the control plane and do not have control on the configuration of the host volume, so there is a risk we may not be able to accept depending on how much impact those backups have.

jewzaam commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+- Have automated backups enabled by default with cluster installation
+- Save cluster backups to PersistentVolumes or cloud storage
+  - This would be a future enhancement or extension to the API
+- Automate cluster restoration
```
Is this expected as a future enhancement?  Not stated here as in the prior bullet.

jewzaam commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+### User Stories
+
+- As a cluster administrator I want to initiate a cluster backup without requiring a root shell on a control plane node so as to minimize the risk involved
+- As a cluster administrator I want to schedule recurring cluster backups so that I have a recent cluster state to recover from in the event of quorum loss (i.e. losing a majority of control-plane nodes)
```
How many backups are retained when there is this schedule?

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+### API Extensions
+
+A new CRD `EtcdBackup` will be introduced to the API group `config.openshift.io` with the version `v1alpha1`. The CRD would be feature-gated until the prerequisite e2e test has an acceptable pass rate. 
```
I think `Backup` with a region for etcd and possibly a future region for other items.  Velero, PVs, somethng else.

deads2k replied:
glad to see alpha here.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+    // The default "" means no backup is triggered
+    // +kubebuilder:default:=""
+    // +optional
+    Reason string `json:"reason"`
```
In our existing operators, we have a `forceRedeploymentReason` or some such, right?  Given your need to track backups, it may be more practical to have an object like the Insights collection that indicates a desire to run a backup "now".

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+    // Default: ""
+    // TODO: Define how we do validation for the format (e.g validation webhooks)
+    // and the limits on the frequency to disallow unrealistic schedules (e.g */2 * * * * every 2 mins)
+    // TODO: What is the behavior if the backup doesn't complete in the specified interval
```
having an object like Insights collection helps here too, since it gives an "obvious" item to delete.  Versus finding a specific instance of a job in a managed namespace.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+      // every day at 3am: 0 3 * * *
+    // Setting to an empty string "" means disabling scheduled backups
+    // Default: ""
+    // TODO: Define how we do validation for the format (e.g validation webhooks)
```
try CEL first

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+    // TODO: Validation on min and max durations.
+        // The max should be smaller than the scheduled interval
+        // for the next backup else the alert will never be generated
+    MaxDurationToCompleteBackup string `json: "maxDurationWithoutBackup"`
```
I have no preference for a timeout as long as the operator reports degraded.

deads2k commented on enhancements/etcd/automated-backups.md:
````diff
@@ -0,0 +1,284 @@
+```
+
+**TODO:** How do we feature gate the entire type?
+For a field within a non-gated type it seems to be the tag [`// +openshift:enable:FeatureSets=TechPreviewNoUpgrade`](https://github.com/openshift/api/blob/master/example/v1/types_stable.go#L33). No such tag visible on the [whole type gated example](https://github.com/openshift/api/blob/1957a8d7445bf2332f027f93a24d7573f77a0dc0/example/v1alpha1/types_notstable.go#L19).
````
example of a gate on an entire type here: https://github.com/openshift/api/blob/master/example/v1alpha1/0000_50_notstabletype-techpreview.crd.yaml

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+    // Conditions represents the observations of the EtcdBackup's current state.
+    // TODO: Identify different condition types/reasons that will be needed.
+    // +optional
```
in the actual PR, you'll need the SSA tags.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+    // EtcdBackup. It corresponds to the EtcdBackup's .metadata.generation that the condition was set based upon.
+    // This lets us know whether the latest condition is stale.
+    // +optional
+    ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```
This hasn't been done in our config API before.  I suggest not starting now.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+    // Conditions represents the observations of the EtcdBackup's current state.
+    // TODO: Identify different condition types/reasons that will be needed.
+    // +optional
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
```
This surprises me.  Having status on configuration is not common.  Try a thought experiment modelling, "backup now" requests as a separate resource like https://github.com/openshift/api/blob/master/insights/v1alpha1/types_insights.go#L16 and see how it fairs.  I think it gives you some greater flexibility and you'll be able to order "one at a time" with a "brush your teeth policy" of only the latest is honored and all others get updated as "won't do".

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+#### Execution method
+
+Create a backup pod that runs the backup script on a designated master node to save the backup file on the node's host filesystem:
```
the job construct seems like a good choice here.  I don't object to a pod, but are you sure you don't want a job instead?

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+  - If significant, should we avoid selecting the leader to reduce the risk of overloading the leader and triggering a leader election?
+  - We can find the leader via the [endpoint status](https://etcd.io/docs/v3.5/dev-guide/api_reference_v3/#message-statusresponse-apietcdserverpbrpcproto) but the leader can change during the course of the backup or we may not have the option of filtering the leader out if the backup pod is unschedulable on non-leader nodes (e.g low disk space). 
+- Other than the leader IO considerations, when picking where to run and save the backup, do we need to load balance or round-robin across all control-plane nodes?
+  - Leaving it to the scheduler to schedule the backup pod (or having the operator save on the node it's running on) is simpler, but we could end up with all/most backups being saved on a single node. In a disaster recovery scenario we could lose that node and the backups saved.
```
I think a force assignment policy of some kind is a good choice.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+### Test Plan
+
+Prior to merging the backup controller implementation, an e2e test would be added to openshift/origin to test the validity of the backups generated from the automated backups feature. The pass rate on this test will feature gate the API until we pass an acceptable threshold to make it available by default.
```
Gating by techpreview means you can merge your impl at will.

deads2k replied:
promoting to "accessible-by-default" will be allowed once there is an acceptable history of passing restore

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,284 @@
+
+#### Tech Preview -> GA
+
+- TBD
```
this is known. Track record of reliable restore.  I propose 99% on all platforms as a starting point since 99% of our tests currently pass at this threshold and consequences of failed restore is extremely high.

hasbro17 commented:
Some updates:
- Removed `MaxDurationToCompleteBackup` alert in favor of simply making the backup controller go degraded if it notices a backup job didn’t complete in some default timeout.
- Replaced RetentionCount int with a discriminated union field `RetentionPolicy`  which supports retention by count or size. Leaves us open to adding more (by time) later.
- Added some CEL validation rules for API spec.
- Clarified that non-self hosted clusters like Hypershift are out of scope for the enhancement
- Added a section on backup timeout failure and setting a degraded condition.
- Graduation criteria states the feature gate and 99% pass rate requirement.
- Expanded the Test Plan section a little bit with how we want to validate a successful restore.

The most significant change however is splitting up the original EtcdBackup API and controller into two different APIs and controllers so we can better track the status for individual backup requests without lumping them into conditions on the config API (https://github.com/openshift/enhancements/pull/1370#discussion_r1149857906):
- `EtcdBackupRequest` for handling one time backup requests so we can better track the status of multiple past backups
- `EtcdBackupSchedule` for configuring the schedule for recurring backups which outsources the actual backup requests via `EtcdBackupRequest`

For now another look at the updated Workflow and API sections would be helpful to know whether the two APIs work better for configuration and status tracking. @deads2k 

I still have some other ToDos to follow up. Primarily https://github.com/openshift/enhancements/pull/1370#discussion_r1143301692 to figure out on how we would enforce the retention policy for backup files spread across all nodes but will update in another follow up.

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -65,101 +67,147 @@ node.
+A new controller in the cluster-etcd-operator, `EtcdBackupRequestController`, will reconcile backup requests as follows:
+- Watch for new `EtcdBackupRequest` CRs as created by an admin
+- Create a backup Job configured for the backup request spec
+- Track the backup progress, failure or success on the `EtcdBackupRequest` status
```
I assume you will also need to track:
* file path
* node name

so we can later sort out the retention, correct?

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -65,101 +67,147 @@ node.
+- Watches the `EtcdBackupSchedule` CR as created by an admin
+- Creates a [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) that would in turn create etcd backup requests `EtcdBackupRequest` at the desired schedule
+- Updates the CronJob for any changes in the schedule
+- Fulfils the specified retention policy by pruning existing backup files before creating a new `EtcdBackupRequest`
```
is the cronjob managing the retention or the controller in CEO?

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -191,9 +239,14 @@ The backup pod needs to be able to save the backup data onto the host filesystem
 
-TBD
+If a backup job fails to complete after a default timeout interval e.g 5 mins the backup controller should set a `BackupControllerDegraded` condition on the `etcds.operator.openshift.io/v1` `cluster` CR. 
+The time to complete a backup should typically be in the range of 30s ( see [upstream FAQ on snapshot timeout](https://etcd.io/docs/v3.5/faq/#what-does-the-etcd-warning-snapshotting-is-taking-more-than-x-seconds-to-finish--mean)). However depending on the size of the data and the hardware this may take longer.
```
I think 5m is safe, even for 5000 node clusters. Do the math: there are 8 gigs of mmap'd file that need to be dumped at max, big clusters usually have fast disks - so that shouldn't take longer than a minute (worst case).

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -191,9 +239,14 @@ The backup pod needs to be able to save the backup data onto the host filesystem
+If a backup job fails to complete after a default timeout interval e.g 5 mins the backup controller should set a `BackupControllerDegraded` condition on the `etcds.operator.openshift.io/v1` `cluster` CR. 
+The time to complete a backup should typically be in the range of 30s ( see [upstream FAQ on snapshot timeout](https://etcd.io/docs/v3.5/faq/#what-does-the-etcd-warning-snapshotting-is-taking-more-than-x-seconds-to-finish--mean)). However depending on the size of the data and the hardware this may take longer.
+
+**TODO:** Should we provide an operator flag for overriding the default backup timeout for large or slow clusters?
```
maybe add that option to the backup request+schedule CRD? I don't think we want to encourage people to set their CEO to unmanaged in order to configure a timeout

tjungblu commented on enhancements/etcd/automated-backups.md:
```diff
@@ -223,6 +276,8 @@ This test would effectively run through the backup and restore procedure as foll
 - Step through the recovery procedure from the saved backup
 - Ensure that the control-plane has recovered and is stable
 - Validate that we have restored the cluster to the previous state so that it does not have the post-backup changes
+  - Ideally want a way to have all clients/operators restart their watches as they may be trying to watch revisions that do not exist post backup. Restarting the individual operators is one solution but it does not scale well for a large number of workloads on the cluster.
```
> but it does not scale well for a large number of workloads on the cluster.

I think it would work, the problem is more: how do we know what is an operator that has a cached API object? 
In that sense, restarting all nodes would be better - but that indeed doesn't scale that well.

hasbro17 commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+
+Since it is an implementation detail with no API impact we can start with utilizing the backup script for simplicity and codify it later.
+
+#### Saving the backup and retention
+
+When deciding how to save the backup files on the control-plane host machine(s) there are two different approaches with different impacts on the visibility of backups, availability, and maintenance burdens of the retention policies.
+
+### Saving across all nodes
+
+One method is to spread out the saved backups on all available control-plane nodes. This would involve:
+- Scanning all control-plane nodes for existing backup files
+- Pruning the backup file from the node with the oldest backup
+- Saving a new backup on the node that has the oldest most-recent-backup
+
+E.g given the following nodes N(x) with backups B(t) where x is the node suffix and t is the backup timestamp:
+- N1[B1, B5], N2[B2, B4], N3[B3, B6]
+- We would select N1 to prune since B1 is the oldest of (B1,B2,B3)
+- We would select N2 as the node to save the latest backup on since B4 is the oldest of (B4,B5,B6).
+
+This approach has the benefit of spreading out backups across nodes which reduces the likelihood of losing the most recent backup in the event of losing access to a node. However it also has the following drawbacks:
+
+- Gathering the state of the backup files is more complicated:
+  - The backup controller could serially schedule a pod on each node to try and gather the backup state. It's unclear how the controller would gather that state (e.g have each pod write to a configmap).
+- Serially scheduling on each node to gather the state increases the time to perform a backup
+- If a node is unhealthy and is unschedulable then the backup may be stalled as we need to gather the state by scheduling on all nodes
+- Providing visibility into the state of the backup files for the admin is non-trivial:
+  - We would require a status object that lists all nodes and the ordered list of backup filenames per node.
+
+### Saving to a local PV
+
+Alternatively the backup files can be saved at a single location on cluster on a [local Persistent Volume](https://kubernetes.io/docs/concepts/storage/volumes/#local). Saving to a single volume for local storage has the following pros as compared to distributing backups across the control-plane nodes:
+
+- Simplifies the retention logic for locally stored backups. This reduces the maintenance burden of having different retention strategies for locally saved backups vs off-cluster (e.g nfs or gcePersistentDisk volumes).
+- Using a PV also grants the flexibility to do lifecycle management of backups e.g what to do with the old backups if a new schedule is created for a different volume.
+- The saved backups state is more visible from a single location by looking at the PV contents vs looking at all nodes to piece the order of saved backups.
+- For local backups if the cluster is unhealthy the automated backups won't be stalled:
+  - If another node besides the one with the local volume is unavailable we can still save the backup.
+  - If that node with the local volume is inaccessible, we can still provision a new volume on another available node and update the spec to save backups to the new location. 
+- Can leverage [volume snapshots](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) or volume cloning to improve availability of saved backups.
```
Added a section debating the merits of whether we want to distribute and prune backups across all nodes, vs save them to a local volume. Depending on what we decide to pick I'll update the API accordingly (e.g specifying a PVC in the spec).
I'm leaning towards the local PV approach but @williamcaban @deads2k @tjungblu please share your thoughts or concerns.

hasbro17 replied:
Seems like thomas made some comments on the commit but github won't show them. Recapping from my notifications here:

```
@tjungblu: that's only available for CSI drivers:

> VolumeSnapshot support is only available for CSI drivers.

I'm not sure if the local storage is based on CSI nowadays? @jsafrane sorry to pull you into that discussion - do you know if that's available?

@jsafrane: LocalVolumes are not based on CSI and do not support snapshot.
Only on SNO clusters we support [LVMO](https://docs.openshift.com/container-platform/4.12/storage/persistent_storage/persistent_storage_local/persistent-storage-using-lvms.html) and LVM snapshots.
```

Seems like volume snapshots won't be possible for local volumes at least. Not sure if there is another way to replicate/clone the volume. Or write to a secondary local volume as an extra backup. If not, then that is a considerable risk to lose the local volume if the node becomes inaccessible.

williamcaban replied:
My main concern is that the local PV object requires a dedicated disk or partition (both of which are not part of the standard OCP requirements) and the local PV also requires for the cluster to be operational to access the data. How can these constraints be overcome?

deads2k replied:
if local volumes work ootb, I don't mind using them over host mounts.  I think you'd still have to save across nodes.  I don't see local volumes a substitute for general PVs if that's a later requirement.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+
+#### One time backups
+
+To enable one-time backup requests via an API, a new CRD, `EtcdBackupRequest` (open to better name suggestions), will be used to trigger one-time backup requests.
```
`EtcdBackup` is probably enough.

deads2k replied:
This is inside operator.openshift.io?

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -65,101 +67,147 @@ node.
+- Create a backup Job configured for the backup request spec
+- Track the backup progress, failure or success on the `EtcdBackupRequest` status
+
+**TODO:** Decide if `EtcdBackupRequest` needs to be cluster-scoped?
```
cluster-scoped.  magic namespaces make behavior unpredictable for users.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+
+#### Scheduled backups
+
+To enable recurring backups a new cluster-scoped singleton CRD `EtcdBackupSchedule` (open to better name suggestions) will be used to specify the backup schedule, retention policy and other related configuration.
```
How about a Backup.config.openshift.io, with a region under spec for etcd

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+
+The spec will be as listed below, while the status will be empty since the status of individual backups will be tracked on the `EtcdBackupRequest` status block.
+
+**TODO**: While config CRs don't normally seem to have a status block, where do we track the cumulative status e.g total number of backups saved?
```
if the operator is handling retention, something on etcd.operator.openshift.io seems ideal.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+    // TODO: Discuss in multiple backups section. What is the behavior if the backup doesn't complete in the specified interval
+        // e.g: every 1hr but the backup takes 2hrs to complete
+        // Wait for the current one to complete?
+    Schedule string `json:"schedule"`
```
most of the fields you can set reasonable, but timezeone is probably required.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+  // The count here is for the total number of backups
+  // +kubebuilder:validation:Minimum=1
+  // +optional
+  RetentionCount int `json: "retentionCount,omitempty"`
```
make a struct containing this int please.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+  // the oldest backup will be removed before a new backup is initiated.
+  // +kubebuilder:validation:Minimum=0.1
+  // +optional
+  RetentionSizeGb float64 `json: "retentionSizeGb,omitempty"`
```
make a struct containing a resource.quantity please.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+    // Reason defines the reason for the most recent backup.
+    // TODO: Do we need to specify a reason? This is more metadata than spec to reconcile.
+    // We could just have a spec-less CR as created by the admin or the EtcdBackupSchedule per the schedule.
+    // The reason could be tagged into an annotation if needed.
```
we don't use annotations for structured fields.  Reason is entirely reasonable to include as optional here.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+
+**TODO:** If load balancing between control-plane nodes is a requirement then it needs to be determined how we can achieve that with the Job or Pod spec.
+- The [`nodeSelector`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) field would let us choose the group of control-plane/master nodes but not round-robin between them.
+- The [`nodeName`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodename) field would let us choose the node
```
since the operator creates a pod, I encourage directly assigning nodeNames.

deads2k commented on enhancements/etcd/automated-backups.md:
```diff
@@ -0,0 +1,376 @@
+If a backup job fails to complete after a default timeout interval e.g 5 mins the backup controller should set a `BackupControllerDegraded` condition on the `etcds.operator.openshift.io/v1` `cluster` CR. 
+The time to complete a backup should typically be in the range of 30s ( see [upstream FAQ on snapshot timeout](https://etcd.io/docs/v3.5/faq/#what-does-the-etcd-warning-snapshotting-is-taking-more-than-x-seconds-to-finish--mean)). However depending on the size of the data and the hardware this may take longer.
+
+**TODO:** Should we provide an operator flag for overriding the default backup timeout for large or slow clusters?
```
I think not.  Perhaps an auto-tune based on how long previous backups took, but I wouldn't allow configuration.

deads2k commented:
API looks like a reasonable point to start an openshift/api PR from.  I'm happy to see the usage of techpreview.

The implementation details look plausible, but I'll defer to the team there.

hasbro17 commented:
/reopen
/remove-lifecycle rotten

Going to update with the changes from the API PR so far.

dhellmann commented:
@hasbro17 This design is linked to a jira ticket that is in progress. Is the design complete? Who's sign-off do you need to merge this PR?

hasbro17 commented:
@dhellmann My apologies, I didn't get a chance to update this. We've had some design changes over on the API PR that we've already merged (https://github.com/openshift/api/pull/1482) and we're almost finishing up the implementation.

So the design for 4.14 is complete but I need to document the API changes back here and update the proposal for the follow ups and future goals. Will ping the reviewers once I have that done.

/remove-lifecycle rotten

dhellmann commented:
(automated message) This pull request is closed with lifecycle/rotten. The associated Jira ticket, ETCD-81, has status "Testing". Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

dhellmann commented:
(automated message) This pull request is closed with lifecycle/rotten. The associated Jira ticket, ETCD-81, has status "Testing". Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

dhellmann commented:
(automated message) This pull request is closed with lifecycle/rotten. The associated Jira ticket, ETCD-81, has status "Testing". Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

dhellmann commented:
(automated message) This pull request is closed with lifecycle/rotten. The associated Jira ticket, ETCD-81, has status "Testing". Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.
